### ISSUE TITLE
Reproducibility of file ordering in MultiprocessDatasetReader

### ISSUE THREAD

**Issue by pxch** [allenai/allennlp#2620]
**Describe the bug**
The order of input files when using MultiprocessDatasetReader seems to be random in different runs.

**To Reproduce**
When I use MultiprocessDatasetReader with a list of input training files (for example, train_000.json to train_099.json), the order of the training files being loaded is different across multiple runs. I'm using the "allennlp train" entry point with the default random seeds.

**Expected behavior**
The order of input files should be the same given the fixed random seeds to achieve maximum reproducibility.

**System (please complete the following information):**
 - OS: Ubuntu 16.04
 - Python version: 3.7.2
 - AllenNLP version: v0.8.2
 - PyTorch version: v1.0.1

**Additional context**
I tried to manually set random seed by "random.seed(13370)" before calling "random.shuffle(shards)" in the _instances method in MultiprocessDatasetReader, and that gives me the same file order every time. So it seems the global random seed is not affecting the MultiprocessDatasetReader for some reason.

**Comment by brendan-ai2** [comment]
When you say "different runs" do you mean from epoch to epoch or entirely different training runs? And how are you setting the random seed? You can fix these in your config: https://github.com/allenai/allennlp/blob/0bbd359f0840637675a4e6cb9f593871390bb87b/allennlp/common/util.py#L193 If you do so, I would expect the ordering to change from epoch to epoch, but the sequence of orderings to be the same each training run. Given that we typically train with stochastic gradient descent (or variants thereof), I believe this is desirable behavior, but please elaborate if you disagree.

**Comment by pxch** [comment]
I mean entirely different training runs, I definitely agree that the file ordering should change from epoch to epoch.
I used the default seeds (13370) and used the MultiprocessDatasetReader as a wrapper over my customized reader. However the file ordering in the one training run differs from the ordering in another training run.

**Comment by matt-gardner** [comment]
Is this just multiprocessing stochasticity? Like, one processor finishes before another, and so you get different ordering? It's really hard to ensure consistent ordering of things in a multiprocessing environment.

**Comment by pxch** [comment]
I thought about that before. But the ordering was still different even when I used only 1 process.

**Comment by pxch** [comment]
I also tried to print random state by calling `random.getstate()` at different places in the Trainer class as well as in the MultiprocessDatasetReader before `random.shuffle(shards)`. Among multiple training runs, all the random states within the Trainer class kept consistent, but the random states in the MultiprocessDatasetReader class were different in each run. So it seems the random state in the MultiprocessDatasetReader is independent of the rest of the training pipeline.

**Comment by brendan-ai2** [comment]
I looked into this more closely and I've been able to repro, but I don't have a fix yet. Pushing on this a bit further today.

**Comment by brendan-ai2** [comment]
Okay, so this is a doozy. Python has a special one-off hack specifically to reseed `random` when spawning a thread. See https://github.com/python/cpython/blob/3c6b436a57893dd1fae4e072768f41a199076252/Lib/multiprocessing/popen_fork.py#L70.

The intention, IIUC, is to ensure that if you spawn several worker threads, then they will all have separate sources of entropy by default. This seems a bit heavy-handed to me, but I understand the motivation. For context, see the numpy issue requesting the same behavior: https://github.com/numpy/numpy/issues/9650.

Tentative fix here: https://github.com/allenai/allennlp/pull/2729

**Comment by brendan-ai2** [comment]
(To address a possible question, the reader is called in a separate process by the `multiprocess_iterator` here: https://github.com/allenai/allennlp/blob/master/allennlp/data/iterators/multiprocess_iterator.py#L115.)

**Comment by pxch** [comment]
Yes, I only observe the irreproducibility when using multiprocess_reader together with multiprocess_iterator. My temporary clumsy workaround is to pass the random.getstate() to each process in multiprocess_iterator to ensure that the seed stays the same (which seems to just undo the python reseeding hack). Switching to numpy.random is a really neat solution! Thanks!

**Comment by brendan-ai2** [comment]
@pxch , should be fixed now. Please let me know whether this helps!
